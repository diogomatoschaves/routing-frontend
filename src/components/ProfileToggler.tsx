import React from 'react'
import styled from 'styled-components'
import { Box } from '../styledComponents'
import { Icon, Label } from 'semantic-ui-react'
import { UpdateState, Geography } from '../types'
import { PETROL_5 } from '../utils/colours'

interface Props {
  geography?: Geography
  geographies?: Array<Geography>
  width?: string
  updateState: UpdateState
  responseOption?: string
  id: string
}

const StyledIcon = styled(Icon)`
  cursor: pointer;
`

const StyledLabel = styled(Label)`
  cursor: pointer;
  &.ui.label {
    background-color: ${PETROL_5};
    color: white;
  }
`

const getNewProfile = (
  geographies: Array<Geography>,
  geography: Geography,
  left: boolean
) => {
  const index = geographies.findIndex(el => el.name === geography.name)
  const lastIndex = geographies.length - 1
  return left
    ? index === 0
      ? geographies[lastIndex]
      : geographies[index - 1]
    : index === lastIndex
    ? geographies[0]
    : geographies[index + 1]
}

const handleClick = (
  geographies: Array<Geography> | undefined,
  geography: Geography | undefined,
  updateState: UpdateState,
  id: string,
  responseOption: string | undefined
) => {
  updateState(
    id,
    geographies && geography
      ? getNewProfile(geographies, geography, true)
      : responseOption === 'normal'
      ? 'traffic'
      : 'normal'
  )
}

const names: any = {
  normal: 'No Traffic',
  traffic: 'Traffic'
}

const ProfileToggler = ({
  geography,
  geographies,
  width,
  updateState,
  responseOption,
  id
}: Props) => {
  return (
    <Box
      direction="row"
      justify="space-around"
      padding="0 0 0 0"
      height="45px"
      width={width}
    >
      <Box
        width="10%"
        onClick={() =>
          handleClick(geographies, geography, updateState, id, responseOption)
        }
      >
        <StyledIcon color="grey" size="large" name={'angle left'} />
      </Box>
      <Box width="60%">
        <StyledLabel onClick={() => geographies && updateState('recenter', true)} size="large">
          {geography
            ? geography.name
            : responseOption
            ? names[responseOption]
            : responseOption}
        </StyledLabel>
      </Box>
      <Box
        width="10%"
        onClick={() =>
          handleClick(geographies, geography, updateState, id, responseOption)
        }
      >
        <StyledIcon color="grey" size="large" name={'angle right'} />
      </Box>
    </Box>
  )
}

export default ProfileToggler
