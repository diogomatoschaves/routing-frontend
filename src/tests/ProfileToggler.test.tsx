import React from 'react'
import TestRenderer from 'react-test-renderer'
import { MemoryRouter } from 'react-router-dom'
import ProfileToggler from '../components/ProfileToggler';

const geographies = [
  {
    name: 'Berlin',
    coords: [13.38408, 52.51721],
    polygon: 'berlin.geojson'
  },
  {
    name: 'Stuttgart',
    coords: [9.033, 48.7111],
    polygon: 'stuttgart.geojson'
  },
  {
    name: 'Immendingen',
    coords: [8.7214, 47.912],
    polygon: 'immendingen.geojson'
  },
  {
    name: 'San JosÃ©',
    coords: [-121.97588, 37.34606],
    polygon: 'sunnyvale.geojson'
  }
]

it('expect to render ProfileToggler component', () => {

  const testInstance = TestRenderer.create(
    <MemoryRouter initialEntries={[ '/' ]}>
      <ProfileToggler 
        geography={geographies[0]}
        geographies={geographies}
        updateState={jest.fn()}
        id='geography'
      />
    </MemoryRouter>
  )
  expect(testInstance.toJSON()).toMatchSnapshot()
}) 
