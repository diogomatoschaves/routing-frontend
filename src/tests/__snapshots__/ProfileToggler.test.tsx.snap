// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render ProfileToggler component 1`] = `
<div
  className="sc-bdVaJa gxFsig"
  direction="row"
  height="45px"
>
  <div
    className="sc-bdVaJa clGjkA"
    onClick={[Function]}
    width="10%"
  >
    <i
      aria-hidden="true"
      className="grey angle left large icon sc-EHOje eQYihx"
      onClick={[Function]}
    />
  </div>
  <div
    className="sc-bdVaJa gNhOwl"
    width="60%"
  >
    <div
      className="ui large label sc-bZQynM cRUavP"
      onClick={[Function]}
    >
      Berlin
    </div>
  </div>
  <div
    className="sc-bdVaJa clGjkA"
    onClick={[Function]}
    width="10%"
  >
    <i
      aria-hidden="true"
      className="grey angle right large icon sc-EHOje eQYihx"
      onClick={[Function]}
    />
  </div>
</div>
`;
